### ISSUE TITLE
Scaffolded Identity "Manage" Layout Refers to Unscaffoldable Layout Page

### ISSUE THREAD

**Issue by anurse** [aspnet/Scaffolding#1225]
_From @pm64 on Wednesday, January 8, 2020 7:04:03 PM_

### Describe the bug
The scaffolded Areas\Identity\Pages\Account\Manage\\_Layout.cshtml page references "/Areas/Identity/Pages/_Layout.cshtml" as its layout. This page is not present in the project template and is not available for scaffolding.

### To Reproduce
1. In VS2019, create a new ASP.Net Core web application using "Individual User Accounts" authentication
2. Right-click project node in Solution Explorer, select Add | New Scaffolded Item...
3. Use the scaffolder to add the Identity page Areas\Identity\Pages\Account\Manage\_Layout.cshtml to the project
4. Observe that the generated _Layout.cshtml references "/Areas/Identity/Pages/_Layout.cshtml" as its layout. This page does not exist in the project template and is not itself available for scaffolding.

### Notes
- A workaround is to change the layout to "/Pages/Shared/_Layout.cshtml", a page that does reside in the project template. But I don't believe this is the intended layout (though it may be very similar), since "Areas\Identity\Pages\Account\Manage\_Layout.cshtml" does get rendered at runtime. The intended layout appears to be imprisoned in the Identity DLL.

### Further technical details
- ASP.NET Core version 3.1
- Output of `dotnet --info`:
.NET Core SDK (reflecting any global.json):
 Version:   3.1.100
 Commit:    cd82f021f4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18363
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.1.100\

Host (useful for support):
  Version: 3.1.0
  Commit:  65f04fb6db

.NET Core SDKs installed:
  3.1.100 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download

- The IDE (VS / VS Code/ VS4Mac) you're running on, and it's version:
Visual Studio Enterprise 2019  v16.4.2

_Copied from original issue: dotnet/aspnetcore#18200_

**Comment by anurse** [comment]
_From @blowdart on Wednesday, January 8, 2020 8:07:49 PM_

@danroth27  This seems to be blazor specific.

**Comment by anurse** [comment]
_From @pm64 on Wednesday, January 8, 2020 8:13:25 PM_

@blowdart @danroth27 I thought so at first, but then confirmed that traditional web apps are also affected. What leads you to this conclusion? It's quite possible I missed something.

**Comment by anurse** [comment]
_From @blowdart on Wednesday, January 8, 2020 8:15:59 PM_

Ah that wasn't clear from the issue, as the one you linked to was blazor and no-one has brought this up before.

**Comment by anurse** [comment]
_From @blowdart on Wednesday, January 8, 2020 8:16:52 PM_

@javiercn Is this a bug, or are we missing something?

**Comment by anurse** [comment]
_From @javiercn on Wednesday, January 8, 2020 8:21:52 PM_

@blowdart That sounds like a bug.

The scaffolder should detect (within reason) whether the user is using the included layout or has overridden it with the one on the page and provide the path to the right layout on the scaffolded code.

(It can determine what layout is being used by the presence of the _ViewImports file in `/Areas/Identity/Pages`)

**Comment by anurse** [comment]
_From @blowdart on Wednesday, January 8, 2020 8:26:04 PM_

@anurse Can you move this to scaffolding repo please?

**Comment by anurse** [comment]
_From @pm64 on Thursday, January 9, 2020 1:52:11 AM_

@blowdart can we remove the area-blazor tag? While Blazor is affected, I don't want this bug to get clumped in with existing known scaffolding issues in Blazor.

**Comment by deepchoudhery** [comment]
@HaoK that is the template that we have in the scaffolding repo. Is this intentional?

**Comment by deepchoudhery** [comment]
@javiercn should the account.manage layout be referencing the default/user defined layout or the aforementioned one?

**Comment by iJoris** [comment]
I can confirm that i get the same issue on the latest VS version.

**Comment by Rick-Anderson** [comment]
@HaoK @javiercn @deepchoudhery  
It's ***not*** Blazor specific, All the templates use code very similar to https://github.com/dotnet/Scaffolding/blob/master/src/VS.Web.CG.Mvc/Templates/Identity/Bootstrap4/Pages/Account/Manage/Account.Manage._Layout.cshtml#L8

There is no such file as `Layout = "/Areas/Identity/Pages/_Layout.cshtml";` but it doesn't matter as `if (ViewData.TryGetValue("ParentLayout", out var parentLayout))` always returns true. At least I can't figure out how  to get it to return false.

![image](https://user-images.githubusercontent.com/3605364/84218546-59ea3d00-aa6a-11ea-8239-4a0ae2a4233d.png)

The code is incorrect,  it should be
```
"/Pages/Shared/_Layout.cshtml";
```
For Razor Pages, and the views folder for MVC.

This is trivial to repo with any of the web  templates. For example:

```
dotnet new webapp -au Individual -uld -o RPauth
cd RPauth
dotnet add package Microsoft.VisualStudio.Web.CodeGeneration.Design
dotnet aspnet-codegenerator identity -dc RPauth.Data.ApplicationDbContext 
```

I could probably PR this if you like.

**Comment by HaoK** [comment]
Yeah its not supposed to match exactly, the UI does some magic to map the area path /Identity to the actual file

**Comment by Rick-Anderson** [comment]
I'll document that.

This issue and  duplicate #931 should be closed.

In the [Scaffold Identity](https://docs.microsoft.com/en-us/aspnet/core/security/authentication/scaffold-identity?view=aspnetcore-3.1&tabs=visual-studio) doc I'll write something like `Layout = "/Areas/Identity/Pages/_Layout.cshtml";` is correct because the user interface maps  `/Identity` to the actual file

**Comment by javiercn** [comment]
The code is correct, we try to find the layout on the Area if there is no layout on the app, but use the layout on the app if present. SPA templates for example don't have a layout on the app and everything is self-contained within the Identity area. Blazor is a similar deal.

**Comment by AndrewTriesToCode** [comment]
FYI for anyone following along this issue seems to be resolved now. I noticed in August this file was added which explicity sets the parent view so that the bad path is not used. [/Areas/Identity/Pages/Account/Manage/_ViewStart.cshtml](https://github.com/dotnet/aspnetcore/blob/main/src/Identity/UI/src/Areas/Identity/Pages/V5/Account/Manage/_ViewStart.cshtml)

What's interesting is that this file doesn't get scaffolded so I spent a little while scratching my head on where `ParentLayout` was set until I checked the source on GitHub.

**Comment by danroth27** [comment]
Verified with the current Visual Studio flow and independently with `dotnet-aspnet-codegenerator` 10.0.2 using the single-file `Account.Manage._Layout` scenario.

The generated Manage layout still contains the `/Areas/Identity/Pages/_Layout.cshtml` fallback, but `ParentLayout` is supplied at runtime by the Identity UI's embedded `Manage/_ViewStart.cshtml`. After registering and signing in, `/Identity/Account/Manage` rendered successfully using the application layout.

The generated reference is therefore intentional and does not produce the reported functional failure. Closing as by design. This behavior is part of the generic Identity UI layout mechanism and is not Blazor-specific.
